internalsdk/android_test.go

```diff
@@ -77,30 +78,50 @@ func (c testSession) IsPlayVersion() (bool, error)                       { retur
 func (c testSession) SetShowGoogleAds(enabled bool)                      {}
 func (c testSession) SetHasConfigFetched(enabled bool)                   {}
 func (c testSession) SetHasProxyFetched(enabled bool)                    {}
-func (c testSession) SetOnSuccess(enabled bool)                          {}
 func (c testSession) ChatEnable() bool                                   { return false }
 
+func (c testSession) SetOnSuccess(enabled bool) {
+	if !enabled {
+		return
+	}
+	// we need to wait for flashlight.fastconnect dialer to "test" and get a successful connection
+	// before any dialer can be used ("no top dialer" error). fastConnect will call onSuccess when
+	// it gets a successful connection, this give us a way to listen for that before we continue.
+	select {
+	case c.rdy <- true:
+	default:
+	}
```

Comment:
Okay it took me a bit of investigation, but I think I understand now.  All you're doing is ensuring you wait for a successful connection before running the test.  I'd suggest the following to make this more clear to future readers:
- Do away with the flashlight-specific parts of this comment.
- Change the signal to closing of a channel.  That is the standard way of signaling readiness.  The code as-written is fragile in that it requires only a single waiting routine and one which is already waiting.  These requirements are confusing.

So this changes to:
```
func (c testSession) SetOnSuccess(enabled bool) {
	if !enabled {
		return
	}
	// Signal readiness on the first connection to a proxy.
	close(c.rdy)
}
```

And then in your test you have:
```
select {
case <-c.rdy:
case <-time.After(4 * time.Second):
    t.Fatal("session not ready in time")
}
```

And honestly the timeout part is optional since `go test` can specify and manage its own timeout (and the stack trace will tell you where the code got hung up).

Reply:
> Okay it took me a bit of investigation, but I think I understand now. All you're doing is ensuring you wait for a successful connection before running the test.

Sorry about that. I guess that wasn't as clear as I thought it was. 

> Change the signal to closing of a channel. That is the standard way of signaling readiness.

That's what I usually do but, in this case, the same `OnSuccess` callback is [called on every successful dial](https://github.com/getlantern/flashlight/blob/main/dialer/fastconnect.go#L90) regardless of the dialer used, so it will panic if `fastconnect` will test more than one. 

>  The code as-written is fragile in that it requires only a single waiting routine and one which is already waiting. These requirements are confusing.

Yeah, it requires `testSession.rdy` to be a buffered chan to avoid blocking `fastconnect`.

Reply:
And I just realized I should use `once.Do(close)`...